java/src/main/native/src/aggregation128_utils.cu

```diff
@@ -0,0 +1,150 @@
+
+  struct stride_functor : public thrust::unary_function<difference_type, difference_type> {
+    __device__ inline difference_type operator()(difference_type i) const { return i * 4; }
+  };
+
+  typedef typename thrust::counting_iterator<difference_type> CountingIterator;
+  typedef typename thrust::transform_iterator<stride_functor, CountingIterator> TransformIterator;
+  typedef typename thrust::permutation_iterator<int32_t const *, TransformIterator>
+      PermutationIterator;
+
+  typedef PermutationIterator iterator;
+
+  chunk_strided_range(int32_t const *start, int32_t const *finish, int chunk_idx)
+      : start(start + chunk_idx), finish(finish + chunk_idx) {}
+
+  iterator begin() const {
+    return PermutationIterator(start, TransformIterator(CountingIterator(0), stride_functor()));
+  }
+
+  iterator end() const { return begin() + ((finish - start) + 3) / 4; }
+
+private:
+  int32_t const *start;
+  int32_t const *finish;
+};
+
+// Functor to reassemble a 128-bit value from four 64-bit chunks with overflow detection.
+class chunk_assembler : public thrust::unary_function<cudf::size_type, __int128_t> {
+public:
+  chunk_assembler(bool *overflows, uint64_t const *chunks0, uint64_t const *chunks1,
+                  uint64_t const *chunks2, int64_t const *chunks3)
+      : overflows(overflows), chunks0(chunks0), chunks1(chunks1), chunks2(chunks2),
+        chunks3(chunks3) {}
+
+  __device__ __int128_t operator()(cudf::size_type i) const {
+    // Starting with the least significant input and moving to the most significant, propagate the
+    // upper 32-bits of the previous column into the next column, i.e.: propagate the "carry" bits
+    // of each 64-bit chunk into the next chunk.
+    uint64_t const c0 = chunks0[i];
+    uint64_t const c1 = chunks1[i] + (c0 >> 32);
+    uint64_t const c2 = chunks2[i] + (c1 >> 32);
+    int64_t const c3 = chunks3[i] + (c2 >> 32);
+    uint64_t const lower64 = (c1 << 32) | static_cast<uint32_t>(c0);
+    int64_t const upper64 = (c3 << 32) | static_cast<uint32_t>(c2);
+
+    // check for overflow by ensuring the sign bit matches the top carry bits
+    int32_t const replicated_sign_bit = static_cast<int32_t>(c3) >> 31;
+    int32_t const top_carry_bits = static_cast<int32_t>(c3 >> 32);
+    overflows[i] = (replicated_sign_bit != top_carry_bits);
+
+    return (static_cast<__int128_t>(upper64) << 64) | lower64;
+  }
+
+private:
+  // output column for overflow detected
+  bool *const overflows;
+
+  // input columns for the four 64-bit values
+  uint64_t const *const chunks0;
+  uint64_t const *const chunks1;
+  uint64_t const *const chunks2;
+  int64_t const *const chunks3;
+};
+
+} // anonymous namespace
+
+namespace cudf::jni {
+
+// Extract a 32-bit chunk from a 128-bit value.
+std::unique_ptr<cudf::column> extract_chunk32(cudf::column_view const &in_col, cudf::data_type type,
+                                              int chunk_idx, rmm::cuda_stream_view stream) {
```

Comment:
Instead of taking a `chunk_idx` here, how about just having this function extract all 4 chunks and return all 4 columns at once? Then you could do the extraction in a single kernel. If you wanted to be extra fancy, you could return a `packed_columns` and do it in a single allocation too.

Reply:
> how about just having this function extract all 4 chunks and return all 4 columns at once?  

I originally had written it that way, but the way we need to use this in Spark expressions requires each chunk to be an expression.  Spark expressions are 1-to-1 mappings, and I need a separate expression for each chunk.  I wouldn't be able to do this without messing around with struct columns and then peeling out the columns from underneath the struct column at the last minute.  Ultimately this was far simpler and still quite fast.